### PR TITLE
Improve getitem tests

### PR DESCRIPTION
**Pull request by petarmhg**
Tests `getitem` on indices with arbitrary (valid) combinations of ints, slices, integer arrays, boolean arrays, etc

### CODE REVIEW

File: tests/tensor_ops/test_getitem.py [hunk 1]
```diff
@@ -220,3 +241,27 @@ def test_getitem_basic_w_adv_fwdprop():
...
+def test_getitem_arbitraryindex_fwdprop():
+    pass
+
+
+@settings(deadline=None)
+@backprop_test_factory(
+    mygrad_func=get_item,
+    true_func=get_item,
+    num_arrays=1,
+    index_to_arr_shapes={0: hnp.array_shapes(max_side=4, max_dims=8)},
```
Comment (rsokl) [inline tests/tensor_ops/test_getitem.py:262]:
```suggestion
    index_to_arr_shapes={0: hnp.array_shapes(min_side=0, max_side=4, min_dims=0, max_dims=5)},
```

File: tests/tensor_ops/test_getitem.py [hunk 2]
```diff
@@ -220,3 +241,27 @@ def test_getitem_basic_w_adv_fwdprop():
 )
 def test_getitem_basic_w_adv_bkprop():
     pass
+
+
+@fwdprop_test_factory(
+    mygrad_func=get_item,
+    true_func=get_item,
+    num_arrays=1,
+    index_to_arr_shapes={0: hnp.array_shapes(max_side=4, max_dims=8)},
```
Comment (rsokl) [inline tests/tensor_ops/test_getitem.py:250]: Note that running this locally revealed the mygrad bug: https://github.com/rsokl/MyGrad/issues/272

```suggestion
    index_to_arr_shapes={0: hnp.array_shapes(min_side=0, max_side=4, min_dims=0, max_dims=5)},
```

File: tests/tensor_ops/test_getitem.py [hunk 3]
```diff
@@ -220,3 +241,27 @@ def test_getitem_basic_w_adv_fwdprop():
...
+    kwargs=dict(index=arb_index_wrap),
+)
+def test_getitem_arbitraryindex_fwdprop():
+    pass
+
+
+@settings(deadline=None)
```
Comment (rsokl) [inline tests/tensor_ops/test_getitem.py:257]: This is a powerful test - it is worth letting hypothesis do a lot of searching for edge cases
```suggestion
@settings(deadline=None, max_examples=500)
```

File: tests/custom_strategies/__init__.py [hunk 4]
```diff
@@ -358,3 +360,158 @@ def valid_shapes(
...
+        Examples
+        --------
+        >>> group_continuous_integers([1, 3, 4, 5, 7, 8])
+        [(1,), (3, 4, 5), (7, 8)]
+        """
+        return [
+            tuple(map(itemgetter(1), g))
+            for k, g in groupby(enumerate(ls), lambda x: x[0] - x[1])
+        ]
+
```
[the remark this answers is not in the document]
Comment (petarmhg) [reply tests/custom_strategies/__init__.py:410]: Ahh nice catch. In the case of empty arrays, I think this fix will make it so that the case

```
np.array([])[[]]
```

could never be checked (though for what reason this would ever need to be used, I don't know). Since the `integer_array_indices` strategy doesn't handle dim-0 axes, I can add a small test to `test_getitem` to check for this

File: tests/tensor_ops/test_getitem.py [hunk 5]
```diff
@@ -48,6 +47,19 @@ def adv_index_bool_wrap(*arrs):
...
+    return arbitrary_indices(arrs[0].shape)
+
+
+def test_index_empty():
+    a = Tensor([])
+    assert a[[]].shape == (0,)
+
+    a.sum().backward()
+    # since a is empty, a.grad should be empty and same shape
+    assert np.allclose(a, a.grad)
```
Comment (rsokl) [inline tests/tensor_ops/test_getitem.py:60]: `np.allclose` isn't strict about shapes with empty arrays

```python
>>> np.allclose(np.array([[]]) , np.array([]))
True
```
Let's check the shapes explicitly

```suggestion
def test_index_empty():
    a = mg.Tensor([])
    b = a[[]]
    
    assert b.shape == (0,)

    b.sum().backward()
    # since a is empty, a.grad should be empty and same shape
    assert a.shape == a.grad.shape
    assert b.shape == b.grad.shape
```

File: tests/tensor_ops/test_getitem.py [hunk 6]
```diff
@@ -48,6 +47,22 @@ def adv_index_bool_wrap(*arrs):
     return hnp.arrays(shape=arrs[0].shape, dtype=bool)
 
 
+def arb_index_wrap(*arrs):
+    return arbitrary_indices(arrs[0].shape)
+
+
+def test_index_empty():
+    a = mg.Tensor([])
```
Comment (rsokl) [inline tests/tensor_ops/test_getitem.py:55]: Ugh sorry. That was my fault.

```suggestion
    a = Tensor([])
```

**Comment by rsokl** [comment]
🥳
